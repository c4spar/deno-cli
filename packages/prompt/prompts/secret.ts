import { encode } from 'https://deno.land/std@v0.52.0/encoding/utf8.ts';
import { blue, green, underline } from 'https://deno.land/std@v0.52.0/fmt/colors.ts';
import { stripeColors } from '../../table/lib/utils.ts';
import { Figures } from '../lib/figures.ts';
import { GenericInput, GenericInputPromptOptions, GenericInputPromptSettings } from '../lib/generic-input.ts';

export interface SecretOptions extends GenericInputPromptOptions<string> {
    hidden?: boolean;
    minLength?: number;
    maxLength?: number;
}

export interface SecretSettings extends GenericInputPromptSettings<string> {
    hidden: boolean;
    minLength: number;
    maxLength: number;
}

export class Secret extends GenericInput<string, SecretSettings> {

    public static async prompt( options: string | SecretOptions ): Promise<string> {

        if ( typeof options === 'string' ) {
            options = { message: options };
        }

        return new this( {
            pointer: blue( Figures.POINTER_SMALL ),
            hidden: false,
            minLength: 1,
            maxLength: Infinity,
            ...options
        } ).prompt();
    }

    protected setPrompt( message: string ) {

        this.screen.cursorHide();

        message += ' ' + this.settings.pointer + ' ';

        const length = encode( stripeColors( message ) ).length;

        const secret = this.settings.hidden ? '' : '*'.repeat( this.input.length );

        message += underline( secret );

        this.write( message );

        this.screen.cursorTo( length - 1 + this.index );
    }

    protected async getSuccessMessage( value: string ) {
        value = this.settings.hidden ? '*'.repeat( 8 ) : '*'.repeat( value.length );
        return `${ await this.getMessage() } ${ this.settings.pointer } ${ green( value ) }`;
    }

    protected validate( value: string ): boolean | string {
        if ( typeof value !== 'string' ) {
            return false;
        }
        if ( value.length < this.settings.minLength ) {
            return `Secret must be longer then ${ this.settings.minLength } but has a length of ${ value.length }.`;
        }
        if ( value.length > this.settings.maxLength ) {
            return `Secret can't be longer then ${ this.settings.maxLength } but has a length of ${ value.length }.`;
        }
        return true;
    }

    protected transform( value: string ): string | undefined {
        return value;
    }

    protected format( value: string ): string {
        return value;
    }
}
