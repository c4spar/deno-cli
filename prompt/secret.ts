import {
  blue,
  bold,
  dim,
  green,
  stripColor,
  underline,
  yellow,
} from "./deps.ts";
import { Figures } from "./figures.ts";
import {
  GenericInput,
  GenericInputKeys,
  GenericInputPromptOptions,
  GenericInputPromptSettings,
} from "./_generic_input.ts";

/** Secret key options. */
export type SecretKeys = GenericInputKeys;

/** Secret prompt options. */
export interface SecretOptions
  extends Omit<GenericInputPromptOptions<string>, "suggestions"> {
  label?: string;
  hidden?: boolean;
  minLength?: number;
  maxLength?: number;
  keys?: SecretKeys;
}

/** Secret prompt settings. */
interface SecretSettings
  extends Omit<GenericInputPromptSettings<string>, "suggestions"> {
  label: string;
  hidden: boolean;
  minLength: number;
  maxLength: number;
  keys?: SecretKeys;
}

/** Secret prompt representation. */
export class Secret extends GenericInput<string, SecretSettings> {
  /** Execute the prompt and show cursor on end. */
  public static prompt(options: string | SecretOptions): Promise<string> {
    if (typeof options === "string") {
      options = { message: options };
    }

    return new this({
      pointer: blue(Figures.POINTER_SMALL),
      label: "Password",
      hidden: false,
      minLength: 0,
      maxLength: Infinity,
      ...options,
    }).prompt();
  }

  protected input(): string {
    return underline(
      this.settings.hidden ? "" : "*".repeat(this.inputValue.length),
    );
  }

  /** Read user input. */
  protected read(): Promise<boolean> {
    if (this.settings.hidden) {
      this.tty.cursorHide();
    }
    return super.read();
  }

  /**
   * Validate input value.
   * @param value User input value.
   * @return True on success, false or error message on error.
   */
  protected validate(value: string): boolean | string {
    if (typeof value !== "string") {
      return false;
    }
    if (value.length < this.settings.minLength) {
      return `${this.settings.label} must be longer then ${this.settings.minLength} but has a length of ${value.length}.`;
    }
    if (value.length > this.settings.maxLength) {
      return `${this.settings.label} can't be longer then ${this.settings.maxLength} but has a length of ${value.length}.`;
    }
    return true;
  }

  /**
   * Map input value to output value.
   * @param value Input value.
   * @return Output value.
   */
  protected transform(value: string): string | undefined {
    return value;
  }

  /**
   * Format output value.
   * @param value Output value.
   */
  protected format(value: string): string {
    return this.settings.hidden ? "*".repeat(8) : "*".repeat(value.length);
  }
}
