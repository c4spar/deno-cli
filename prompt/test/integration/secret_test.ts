import { Secret } from "../../secret.ts";
import { assertSnapshotCall } from "../../../testing/assert_snapshot_call.ts";

await assertSnapshotCall({
  name: "secret prompt",
  meta: import.meta,
  osSuffix: ["windows"],
  steps: {
    "should enter a secret": { stdin: ["123", "\n"] },
  },
  async fn() {
    await Secret.prompt({
      message: "enter your secret",
    });
  },
});
