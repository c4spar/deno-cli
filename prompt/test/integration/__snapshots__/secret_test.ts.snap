export const snapshot = {};

snapshot[`secret prompt > should enter a secret 1`] = `
stdout:
--------------------------
? enter your secret â€º [23G[G[0J
? enter your secret â€º ***[26G[G[0J
? enter your secret â€º ***
[?25h[?25h
--------------------------
stderr:
--------------------------
--------------------------
`;
